### ISSUE TITLE
core data types: symbol?

### ISSUE THREAD

**Issue by dckc** [ocapn/ocapn#46]
In https://github.com/ocapn/ocapn/issues/5#issuecomment-1550673552 @erights writes:
> After some internal discussion, I'd like to float the idea that we omit symbols completely from the data model.

cc @zenhack @cwebber @jar398

**Comment by cwebber** [comment]
To quote the rest of what @erights said:

> If others require symbols, that's fine, we can keep them (and Agoric will deal somehow with the well-known symbol vs registered symbol mess). But don't include them on our account!
>
> Who needs symbols in the protocol? What do symbols in the protocol do for you that just having strings in the protocol do not.

I feel strongly they should be included.  The short version is: strings represent textual information, symbols represent symbolic identifiers of a concept within a program.

Of course one could say: strings and symbols are very similar.  One could say there is no reason to include symbols since we have strings, aren't strings enough?  On that note, there is no need at all to actually have anything in the protocol that is not just tagged lists and bytestrings: it was enough for SPKI/SDSI style rivest s-expressions, which can technically encode everything we can imagine, and that could be enough for us.  That is not the path we seem to be considering however, certainly not if we are at the point of considering [two bottom types](https://github.com/ocapn/ocapn/issues/50).

**Comment by zenhack** [comment]
(Just to clarify my thumbs up: I don't really have any preferences here, I'm fine with whatever consensus others come to).

**Comment by dckc** [comment]
So we have:
 - Spritely: supported, strongly desired
 - Agoric: poorly supported, mild preference to remove, can probably deal with leaving them in somehow
 - capnproto: supported, whatever

That adds up to "they're in", unless we want to set the bar higher about requirements tracing... I could ask @cwebber to tell me a story where demoting symbols to some ancillary position such as a tagged thing would cause trouble. But I'm not inclined to.

So again, this is resolved to my satisfaction: strings and symbols are disjoint subsets of ocapn values.

Again, https://github.com/ocapn/ocapn/issues/3#issuecomment-1552187671 is an adequate specification:

```capnp
# ocapn.capnp
@0xcd301da1d95b8242;

struct Value {
  union {
    ## Atoms ##

...
    string @6 :Text;
    byteString @7 :Data;
    symbol @8 :Text;
```

Until we have a habit of reviewing recently closed issues, I'll make another explicit agenda request: @jar398 please consider this for the agenda of the May meeting #43

**Comment by kriskowal** [comment]
I would like to reöpen this issue. Whether to use symbols or strings under various circumstances will impede interoperation between JavaScript and Guile implementations.

> I feel strongly they should be included. The short version is: strings represent textual information, symbols represent symbolic identifiers of a concept within a program.

This does not match the expectations in JavaScript and will inhibit the creation of cross-language protocols like what @erights would like to produce for `Symbol.for('getInterfaceOf')`.

JavaScript uses symbols and interned symbols to create separate layers of names for API authors and platform authors. They also allow us to mark objects to overcome prototype identity discontinuity or “eval twin” problems (h/t @erights). That is, if the platform libraries are instantiated multiple types, objects marked by symbols are often our only recourse for recognizing objects created by our eval twin. This is relevant even for our CapTP marshaling layer that uses this kind of marker to distinguish remotable objects.

To that end, Agoric will effectively be reserving certain symbols and objects using symbols of the same name.

Additionally, we are interested in establishing protocols like `getInterfaceOf` that we would like to either by named by symbol or string regardless of implementation language. I believe that string is the more likely common ground on the wire.

**Comment by kriskowal** [comment]
A slightly less rushed rationale:

It seems to me that for broad success to occur, we will need not just OCapN, but some protocols that stand on top of OCapN. For example, we already see a protocol for a bootstrap object that implements `fetch` and other methods. I’ve implemented a similar protocol for Endo Pet Daemon, and these are likely to need to converge.

Given that Goblins will favor symbol-named methods and Agoric’s Endo will favor string-named methods, and given that OCapN supports both, we will need to choose one or the other. My position is that this will be a non-issue if the wire protocol omits symbols, such that Goblins uses interned symbols to represent method names and serializes them as strings, whereas Agoric’s Endo will use stringly named methods and serialize them as strings.

It is simultaneously true that JavaScript objects implement both symbolically and stringly named methods, and that these are namespaces reserved for different parties. Most method names are strings and are suitable for representation on the wire. Symbolically named methods are used by platform and library authors to avoid collisions with API methods, for a variety of other protocols like iteration. I expect most of these symbolic protocols in JavaScript to be very JavaScript specific, like the iterator protocol, and for these symbols to be exposed to the wire to reduce round-trip portability.

**Comment by kriskowal** [comment]
Further, in Agoric’s Endo, @erights is considering whether to expose a method to the wire, either the string `getInterfaceOf` or the interned symbol `Symbol.for('getInterfaceOf')`. If we use the symbol, this will not collide with the API authors namespace in JavaScript, but will collide with the API author’s namespace in Goblins.

**Comment by kriskowal** [comment]
It occurs to me that there is a position where O’Cap’n would have “symbols” that is consistent with my position that JavaScript and Lisp should look the same through an OCapN lens with regard to method names and that all method names should either be strings or symbols:

Straw person:
- O’Cap’n uses O’Cap’n symbols exclusively for method names and excludes symbols from passable and remotable values.
- JavaScript reïfies method names as JavaScript strings.
- Lisp reïfies method names as Lisp symbols. Nobody, not even Lisp, can call a symbolically named JavaScript method.
- Nobody, not even JavaScript, can call a stringly named Lisp method.
- No actors in any language can implement a heterogenous combination of symbolically and stringly named methods.

Would that be satisfying?

**Comment by kriskowal** [comment]
Straw person 2 extends Straw person 1 except:

- O’Cap’n supports passable symbols
- JavaScript reïfies O’Cap’n symbols like `Symbol.for('ocapn:method')` such that `E(actor).method()` is equivalent to `E(actor)[Symbol.for('ocapn:method')]()` but `E(actor)[Symbol.for('method')]()` is immediately rejected.

**Comment by dckc** [comment]
Some weekend hacking / brainstorming got me wondering about method invocation compatibility between Agoric and Spritely.
(_I wonder if #65 is the more relevant issue; but this one seems more active..._)

In particular, if I want to integrate Spritely petName work with Agoric stuff, are the namehub APIs compatible? Or do I have to add a wrapper? What would `E(kit.query).lookup('agoricNames', 'brand', 'IST')` look like in spritely / scheme idioms?

I should be able to look this up myself, but any help is appreciated.

background:

 - https://github.com/agoric-labs/gimix/issues/17
 - https://github.com/Agoric/agoric-sdk/discussions/8551
 - https://github.com/dckc/finquick/pull/64/files#diff-ea280ce329c8b9213f1b80075f5591705eb71ec0fc1896ef7502695591335e0eR15

**Comment by dpwiz** [comment]
2 Haskell cents. While the language core/base lacks a Symbol type, there is a familiar syntax to produce values like that:
- `#likeSymbol`
- `"like string"`

So, we're on board, UX-wise (:

**Comment by kriskowal** [comment]
At the September 10, 2024 convocation of the OCapN working group, we achieved consensus that Guile symbols would correspond to OCapN "selectors" and *not* JavaScript registered symbols. We settled on a [calling convention](https://github.com/ocapn/ocapn/issues/131) that would leave only one OCapN operator for delivering both function application and method invocation, and that method invocation would be equivalent to delivering a function application where the first argument is a selector for the method name. A JavaScript would allow the receiver to be implemented either as a function or an object, and if it is an object, would invoke the corresponding string-named method. Although such an object would be unable to receive both selectors and strings for method names, reimplementing such an object as a function can sense that distinction.

I close this issue in my role as editor and the `Meeting` label to ensure this action is reviewed at the October meeting.
